### ISSUE TITLE
I get "anyio._backends._asyncio.ExceptionGroup: 0 exceptions were raised in the task group"

### ISSUE THREAD

**Issue by frederikaalund** [agronholm/anyio#281]
The following code reproduces the issue. Sorry about the length of it.

```python
from __future__ import annotations

from contextlib import suppress
from pathlib import Path
from subprocess import STDOUT, CalledProcessError
from typing import Optional, Sequence, Union

import anyio
from anyio.abc import Process
from anyio.streams.file import FileReadStream
from anyio.streams.text import TextReceiveStream


async def run_process(
    # tg: TaskGroup,
    command: Union[str, Sequence[str]],
    *,
    input_for_stdin: Optional[Path] = None,
    raise_on_rc: Optional[bool] = None,
) -> None:
    """Run the given command as a foreground process.

    Unlike `anyio.run_process`, this streams data to/from the process while the
    process runs. This way, you can see the process' output while it's running.
    Useful for long-running processes.
    """
    process: Optional[Process] = None
    try:
        process = await anyio.open_process(command, stderr=STDOUT)
        await drain_streams(process, input_for_stdin)
    except BaseException:
        if process is not None:
            # Try to gracefully terminate the process
            process.terminate()
            # Give the process some time to stop
            with anyio.move_on_after(5, shield=True):
                await drain_streams(process)
        raise
    finally:
        if process is not None:
            # We tried to be graceful. Now there is no mercy.
            with suppress(ProcessLookupError):
                process.kill()
            # Close the streams (stdin, stdout, stderr)
            await process.aclose()

    assert process.returncode is not None
    # Check the return code (rc)
    if raise_on_rc and process.returncode != 0:
        raise CalledProcessError(process.returncode, command)


async def drain_streams(
    process: Process, input_for_stdin: Optional[Path] = None
) -> None:
    async with anyio.create_task_group() as tg:
        # In parallel:
        #  * send to stdin
        #  * receive from stdout
        if process.stdin is not None and input_for_stdin is not None:
            tg.start_soon(_send_to_stdin, process, input_for_stdin)
        if process.stdout is not None:
            tg.start_soon(_receive_from_stdout, process)
        # Wait for normal exit
        await process.wait()


async def _send_to_stdin(process: Process, input_for_stdin: Path) -> None:
    assert process.stdin is not None
    # Forward data from file to stdin
    async with await FileReadStream.from_path(input_for_stdin) as chunks:
        async for chunk in chunks:
            await process.stdin.send(chunk)


async def _receive_from_stdout(process: Process) -> None:
    assert process.stdout is not None
    # Forward data from stdout
    async for string in TextReceiveStream(process.stdout):
        print(string)


async def main():
    async with anyio.create_task_group() as tg:
        # Run the process in the "background"
        tg.start_soon(run_process, ("sleep", "10"))
        # We can do something else while the process runs
        print("Sleeping now. Try to press CTRL+C.")
        await anyio.sleep(10)


anyio.run(main)
```

Try to press CTRL+C while it runs. Example stack trace:

```
Sleeping now. Try to press CTRL+C.
^Cunhandled exception during asyncio.run() shutdown
task: <Task finished name='__main__.run_process' coro=<run_process() done, defined at /projects/stork/anyio_bug.py:14> exception=<ExceptionGroup: >>
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1854, in _run_once
    event_list = self._selector.select(timeout)
  File "/usr/local/lib/python3.9/selectors.py", line 469, in select
    fd_event_list = self._selector.poll(timeout, max_ev)
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/projects/stork/anyio_bug.py", line 65, in drain_streams
    await process.wait()
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 825, in wait
    return await self._process.wait()
  File "/usr/local/lib/python3.9/asyncio/subprocess.py", line 135, in wait
    return await self._transport._wait()
  File "/usr/local/lib/python3.9/asyncio/base_subprocess.py", line 235, in _wait
    return await waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/projects/stork/anyio_bug.py", line 30, in run_process
    await drain_streams(process, input_for_stdin)
  File "/projects/stork/anyio_bug.py", line 65, in drain_streams
    await process.wait()
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 526, in __aexit__
    raise ExceptionGroup(exceptions)
anyio._backends._asyncio.ExceptionGroup: 0 exceptions were raised in the task group:
----------------------------

Traceback (most recent call last):
  File "/projects/stork/anyio_bug.py", line 92, in <module>
    anyio.run(main)
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_core/_eventloop.py", line 55, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 211, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1854, in _run_once
    event_list = self._selector.select(timeout)
  File "/usr/local/lib/python3.9/selectors.py", line 469, in select
    fd_event_list = self._selector.poll(timeout, max_ev)
KeyboardInterrupt
```

The interesting part to me is the `anyio._backends._asyncio.ExceptionGroup: 0 exceptions were raised in the task group` message. Why does anyio raise an `ExceptionGroup` with zero exceptions in it?

If I'm doing something that I'm not supposed to, let me know. :)

**Comment by agronholm** [comment]
It doesn't – it's just that the list of exceptions gets overwritten. The only case where I've run into this was when task groups were misused, but I don't see that happening here. Is there any way to simplify this code?

You're running the latest AnyIO, yes?

**Comment by frederikaalund** [comment]
Thanks for the quick reply.

> You're running the latest AnyIO, yes?

I was on a commit just-before-3.0.0. I just tried to run it again with the latest master (3a7b67ac269022b0beca06fd8074459db7cf58b4). It gave the same output.

> Is there any way to simplify this code?

It's a lot of code, I know. :) I already reduced it a lot from my production code. I'll try to simplify it further. Hopefully, I can get all the process stuff out of it.

**Comment by frederikaalund** [comment]
If I comment out the "graceful termination" part, the issue doesn't occur:

```python
    except BaseException:
        if process is not None:
            # Try to gracefully terminate the process
            process.terminate()
            # Give the process some time to stop
            # with anyio.move_on_after(5, shield=True):
            # await drain_streams(process)
        raise
```

The problem is that I need that part of the code. :))

**Comment by agronholm** [comment]
What is the exception being caught here?

**Comment by frederikaalund** [comment]
I managed to simplify the example code quite a lot:

```python
import anyio


async def run_process() -> None:
    try:
        await drain_streams()
    except anyio.get_cancelled_exc_class():
        await drain_streams()
        raise


async def drain_streams() -> None:
    async with anyio.create_task_group() as tg:
        tg.start_soon(anyio.sleep, 5)
        await anyio.sleep(5)


async def main() -> None:
    async with anyio.create_task_group() as tg:
        tg.start_soon(run_process)
        print("Sleeping now. Try to press CTRL+C.")
        await anyio.sleep(5)


anyio.run(main)
```

Hope that helps.

> What is the exception being caught here?

It's `anyio.get_cancelled_exc_class()`. I just used BaseException before since I was unsure about `get_cancelled_exc_class` (and I re-raised the exception anyhow).

**Comment by agronholm** [comment]
I'm not able to reproduce this. Which Python version are you running on? Do you have uvloop installed?

Py3.7, with uvloop:
```
$ python bug.py 
Sleeping now. Try to press CTRL+C.
^CTraceback (most recent call last):
  File "bug.py", line 25, in <module>
    anyio.run(main)
  File "/home/alex/workspace/anyio/src/anyio/_core/_eventloop.py", line 55, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/home/alex/workspace/anyio/src/anyio/_backends/_asyncio.py", line 232, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/lib64/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "uvloop/loop.pyx", line 1450, in uvloop.loop.Loop.run_until_complete
  File "uvloop/loop.pyx", line 1443, in uvloop.loop.Loop.run_until_complete
  File "uvloop/loop.pyx", line 1351, in uvloop.loop.Loop.run_forever
  File "uvloop/loop.pyx", line 519, in uvloop.loop.Loop._run
  File "uvloop/handles/poll.pyx", line 213, in uvloop.loop.__on_uvpoll_event
  File "uvloop/cbhandles.pyx", line 90, in uvloop.loop.Handle._run
  File "uvloop/cbhandles.pyx", line 73, in uvloop.loop.Handle._run
  File "uvloop/loop.pyx", line 359, in uvloop.loop.Loop._read_from_self
  File "uvloop/loop.pyx", line 364, in uvloop.loop.Loop._invoke_signals
  File "uvloop/loop.pyx", line 339, in uvloop.loop.Loop._ceval_process_signals
KeyboardInterrupt
```
No uvloop:
```
$ python bug.py 
Sleeping now. Try to press CTRL+C.
^CTraceback (most recent call last):
  File "bug.py", line 25, in <module>
    anyio.run(main, backend_options={'use_uvloop': False})
  File "/home/alex/workspace/anyio/src/anyio/_core/_eventloop.py", line 55, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/home/alex/workspace/anyio/src/anyio/_backends/_asyncio.py", line 232, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/lib64/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.7/asyncio/base_events.py", line 574, in run_until_complete
    self.run_forever()
  File "/usr/lib64/python3.7/asyncio/base_events.py", line 541, in run_forever
    self._run_once()
  File "/usr/lib64/python3.7/asyncio/base_events.py", line 1750, in _run_once
    event_list = self._selector.select(timeout)
  File "/usr/lib64/python3.7/selectors.py", line 468, in select
    fd_event_list = self._selector.poll(timeout, max_ev)
KeyboardInterrupt
```

**Comment by agronholm** [comment]
Ah-ha, py3.8 and up reproduce the problem.

**Comment by frederikaalund** [comment]
I'm on python `Python 3.9.2`.

Even simpler repro example:

```python
import anyio


async def drain_streams() -> None:
    async with anyio.create_task_group() as tg:
        tg.start_soon(anyio.sleep, 5)
        await anyio.sleep(5)


async def main() -> None:
    async with anyio.create_task_group() as tg:
        tg.start_soon(drain_streams)
        print("Sleeping now. Try to press CTRL+C.")
        await anyio.sleep(5)


anyio.run(main)


```

**Comment by agronholm** [comment]
Thanks for simplifying the repro code. I think I can take it from here.

**Comment by frederikaalund** [comment]
Thanks for the quick response. Let me know if there's anything I can do.

Last simplification:

```python
import anyio


async def raise_exception() -> None:
    raise KeyboardInterrupt()


async def main() -> None:
    async with anyio.create_task_group() as tg:
        tg.start_soon(raise_exception)


anyio.run(main)
```

Outputs:

```
Traceback (most recent call last):
  File "/projects/stork/anyio_bug.py", line 13, in <module>
    anyio.run(main)
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_core/_eventloop.py", line 55, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 232, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 47, in run
    _cancel_all_tasks(loop)
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 63, in _cancel_all_tasks
    loop.run_until_complete(
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
    handle._run()
  File "/usr/local/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 227, in wrapper
    return await func(*args)
  File "/projects/stork/anyio_bug.py", line 10, in main
    tg.start_soon(raise_exception)
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 545, in __aexit__
    raise exceptions[0]
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
    handle._run()
  File "/usr/local/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/projects/stork/anyio_bug.py", line 5, in raise_exception
    raise KeyboardInterrupt()
KeyboardInterrupt
Task exception was never retrieved
future: <Task finished name='__main__.main' coro=<main() done, defined at /projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py:218> exception=KeyboardInterrupt()>
Traceback (most recent call last):
  File "/projects/stork/anyio_bug.py", line 13, in <module>
    anyio.run(main)
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_core/_eventloop.py", line 55, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 232, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 47, in run
    _cancel_all_tasks(loop)
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 63, in _cancel_all_tasks
    loop.run_until_complete(
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
    handle._run()
  File "/usr/local/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 227, in wrapper
    return await func(*args)
  File "/projects/stork/anyio_bug.py", line 10, in main
    tg.start_soon(raise_exception)
  File "/projects/stork/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 545, in __aexit__
    raise exceptions[0]
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
    handle._run()
  File "/usr/local/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/projects/stork/anyio_bug.py", line 5, in raise_exception
    raise KeyboardInterrupt()
KeyboardInterrupt
```

**Comment by agronholm** [comment]
The output does not have anything to do with exception groups though, and that was the issue here, yes?

**Comment by frederikaalund** [comment]
You're right.

I just noticed the `Task exception was never retrieved`. Is this intended or a separate issue?

**Comment by agronholm** [comment]
> I just noticed the `Task exception was never retrieved`. Is this intended or a separate issue?

It's an asyncio thing, out of AnyIO's control.

**Comment by agronholm** [comment]
I remember investigating this at some point and I eventually did figure out what caused this. That exception group should have had two exceptions in it, both `CancelledError`. Something causes the exceptions to go poof, but I can't remember what that was. This should be fixed nonetheless as it looks pretty odd.

**Comment by frederikaalund** [comment]
I think you're spot on. I actually *did* do some investigation myself and also found that it raised an `ExceptionGroup` with two `CancelledError`s (logged in `TaskGroup.__aexit__`). Later (in `ExceptionGroup.__str__`), both exceptions were gone. I stopped there since I couldn't make sense of it and posted the issue here instead.

**Comment by agronholm** [comment]
The line where the exceptions list is replaced can be found here: https://github.com/agronholm/anyio/blob/master/src/anyio/_backends/_asyncio.py#L559

It's not clear if the logic for when to filter out `CancelledError` is correct. Perhaps we should be checking if the host task has been cancelled. Either way, I believe I can fix this one particular wart without any side effects.

**Comment by agronholm** [comment]
We discussed this on Gitter and a solution came up: if all the exceptions are cancellation errors, just raise one (asyncio knows to ignore this if it one propagates out of a cancelled task).

**Comment by frederikaalund** [comment]
I think that sounds like a good and pragmatic solution.

Does this fix:
 * avoid the whole `unhandled exception during asyncio.run() shutdown` situation to begin with?
 * simply change the stack trace from  `anyio._backends._asyncio.ExceptionGroup: 0 exceptions were raised in the task group` to `asyncio.exceptions.CancelledError`?

I'd prefer the former but maybe that's impossible with the asyncio backend.

**Comment by graingert** [comment]
> I think that sounds like a good and pragmatic solution.
> 
> Does this fix:
> 
>     * avoid the whole `unhandled exception during asyncio.run() shutdown` situation to begin with?

yes, asyncio.run ignores CancelledError exceptions here https://github.com/python/cpython/blob/ce47addfb6f176fad053431b537b77a5f170765e/Lib/asyncio/runners.py#L67-L68

>     * simply change the stack trace from  `anyio._backends._asyncio.ExceptionGroup: 0 exceptions were raised in the task group` to `asyncio.exceptions.CancelledError`?

there won't be a an ExceptionGroup as long as all the exceptions are CancelledError

> 
> I'd prefer the former but maybe that's impossible with the asyncio backend.
